web_api.md

```diff
@@ -0,0 +1,318 @@
+Server via a [configuration file](#processing-server). However, if it is necessary to do so, there are two ways to start
+a Processing Worker:
+
+```shell
+# 1. Use ocrd CLI bundled with OCR-D/core
+$ ocrd processing-worker <processor-name> --queue=<queue-address> --database=<database-address>
+
+# 2. Use processor name
+<processor-name> processing-worker --queue=<queue-address> --database=<database-address>
+```
+
+* `--queue`: a [Rabbit MQ connection string](https://www.rabbitmq.com/uri-spec.html) to a running instance.
+* `--database`: a [MongoDB connection string](https://www.mongodb.com/docs/manual/reference/connection-string/) to a
+  running instance.
+
+### Process Queue
+
+By using a queuing system, specifically [RabbitMQ](https://www.rabbitmq.com/), the reliability and flexibility of the
+whole system are greatly improved. In our implementation, manual acknowledgement mode is used. This means, when a
+Processing Worker finishes successfully, it sends a positive ACK signal to RabbitMQ. In case of failure, it tries again
+three times before sending a negative ACK signal. When a negative signal is received, RabbitMQ will re-queue the
+message. If there is not any ACK signal sent for any reason (e.g. consumer crash, power outage, network problem, etc.),
+RabbitMQ will automatically re-queue the message after timeout, which is 30 minutes by default. This behavior, however,
+can be [overridden](https://www.rabbitmq.com/consumers.html#acknowledgement-timeout) by setting another value for
+the `consumer_timeout` property in the [`rabbitmq.conf`](https://www.rabbitmq.com/configure.html#config-file) file.
+
+To avoid processing the same input twice (in case of re-queuing), a Processing Worker first checks
+the [`redeliver`](https://www.rabbitmq.com/confirms.html#automatic-requeueing) property to see if this message was
+re-queued. If yes, and the status of this process in the database is not `SUCCESS`, it will process the data described
+in the message again.
+
+When a Processing Server receives a request, it creates a message based on the request content, then push it to an
+appropriate queue. A processing queue always has the same name as its respective processors. For
+example, `ocrd-olena-binarize`processors listen only to the queue named `ocrd-olena-binarize`. Below is an example of
+how a message looks like. For a detailed schema, please check
+the [message schema](web_api/processing-message.schema.yml).
+
+```yaml
+job_id: uuid
+processor_name: ocrd-cis-ocropy-binarize
+
+path_to_mets: /path/to/mets.xml
+input_file_grps:
+  - OCR-D-DEFAULT
+output_file_grps:
+  - OCR-D-BIN
+page_id: PHYS_001,PHYS_002
+parameters:
+  params_1: 1
+  params_2: 2
+result_queue_name: ocrd-cis-ocropy-binarize-result
+
+created_time: 1668782988590
+```
+
+In the message content, except `job_id`, `processor_name`, and `created_time`, are added by the Processing Server. The
+rest comes from the body of the `POST /processor/{executable}` request.
```

Comment:
I cannot decode this logically. Content minus job_id, or minus job_id and processor_name and created_time?

Reply:
Sorry, it should be a comma after "Processing Server", but somehow I put a dot. I will re-formulate it like this:

> In the message content, `job_id`, `processor_name`, and `created_time`, are added by the Processing Server, and the rest comes from the body of the `POST /processor/{executable}` request.